README.md

```diff
@@ -1,2 +1,159 @@
+run the program passed in as a parameter. The problem with that is:
+
+1. It's an I/O blocking operation
+2. Once triggered the program will run, there is no way back. That's equivalent
+ of droping down a pebble at the top of a cliff
+
+What I propose is to create 2 new syscalls:
+
+- `prewarm`: which will do pretty much everything that `execve` does however, it will
+   set a new task state to [`TASK_NEW`](https://elixir.bootlin.com/linux/v6.1.1/source/include/linux/sched.h#L100)
+   on the OS scheduler so the process is never scheduled for execution until `runpid`
+   syscall is triggered. (This would leverage the deferred-executions in the kernel to make it
+   as async as possible).
+
+- `runpid`: which will change existing task state to [`TASK_RUNNING`](https://elixir.bootlin.com/linux/v6.1.1/source/include/linux/sched.h#L85)
+   which essentially tells the OS scheduler that it can run at scheduler's earliest
+   convinience.
+
+PS: names of syscalls are most definitely open for suggestions, I struggle with giving 
+things code names. I'm horrible at this :-)
+
+## user space - bash interpreter
+
+My idea is to add a `shopt` option that once enabled, it would instruct bash interpreter
+that upload script loading, it would:
+
+1. extract all commands to be sent to the OS from its parser grammar. Bash already parses it's
+   contents upon execution but for different purposes as of now. If you want to take a look at how
+   that works, I suggest the `parse.y`file - that's a grammar file crafted for the famous YACC parser
+2. make a syscall `prewarm` for every program in the script ahead of time and collect their PIDs.
+3. when it's time to actually run the program, it will make a syscall `runpid` specifying the PID of the
+   warmed up process.
+   
+So from the user's perspective, it would like something like this:
+
+```bash
+#!/bin/bash
+
+# enable prewarm pids mechanism. When not specified, 
+# bash would work as it normally does (through execve syscalls)
+shopt -s prewarm_pids
+
+echo "test" > /tmp/file.txt
+cat /tmp/file.txt
+rm -f /tmp/file.txt
+```
+
+## important notes
+
+There are many ways to implement this in kernel space and I'm not 100% yet which one will be 
+the easiest one / most effective one... so to avoid getting ahead of myself I listed only the
+core points with the simplest implementation (new syscalls) but I'm aware that it's incredibly
+hard get kernel developers to agree with a new syscall :-) So different approches will have
+to be investigated too until we settle on how to make it work.
+
+As for the Bash interpreter, I'm not too set on Bash but given that this is still arguably
+the most used shell interpreter and the licenses being compatible, it may be the best candidate
+for now.
+
+## what I'm looking for out this PoC
+
+1. I first want to make sure that the performance gain is relevant to the point that
+   other people would care about (something greater than 30%). So at first I'm more interested
```

Comment:
+1 quick and dirty. Is this possible to mock without hacking on the kernel? Or if you can lead that, probably it's OK (I don't know how to work on the kernel but am interested to learn).

Reply:
kernel dev is right up my alley. I can lead the kernel implementation and you lead the user-space part (whether it's bash or nushell). How does that sound to you? (do you suggest any different arrangement - I'm flexible with that)